### DISCUSSION TITLE
real time update mechanism for the chart

### DISCUSSION THREAD

**Discussion by yezhengli-Mr9** [GSI-CS-CO/chart-fx#647] [General]
For #641, try to figure out **real time update mechanism for the chart** suitable for 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-chart/src/main/java/io/fair_acc/chartfx/renderer/spi/financial/CandleStickRenderer.java#L105

>  (1) real time candlestick Chart with lines and points together

 for example, 
(1a) incremental update?
(1b) need to integrate with real-time streaming data.
(2) Do not feel these financial examples are (updated) real time since they do not update the similar to 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/chart/Histogram2DimSample.java#L138-L144
likely because these "real-time" financial examples are not handling streaming(-like) data but calling `loadTestData(...)` instead
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/AbstractBasicFinancialApplication.java#L207-L215

Need real-time update of daily minute-level financial charts.

**Comment by yezhengli-Mr9** [comment]
By the way, to accomplish #641, I was able to conduct substantial modification on several classes relevant to 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-dataset/src/main/java/io/fair_acc/dataset/spi/financial/OhlcvDataSet.java#L19
which seems enough to resolve issue (1) and hopefully can resolve (2) as well in the near future.

Understand real-time update mechanism seems one of the very few (two~three) bottlenecks remaining to fully finish #641

**Comment by raven2cz** [comment]
Class `FinancialRealtimeCandlestickSample` is example for realtime ohlcv stream processing with simple execution platform. 

```java
new SimpleOhlcvDailyParser().getContinuousOHLCV(data);
// look to this method
convertTsRowStream() 
// and
DefaultOHLCV
```

Next important classes for you for realtime processing of ohlcv data.

```java
SimpleOhlcvReplayDataSet
SCIDByNio.createTickDataReplayStream()
// creates ohlcv tick data processing provider
TickOhlcvDataProvider
de.gsi.financial.samples.service.SimpleOhlcvReplayDataSet#fillTestData
de.gsi.financial.samples.service.SimpleOhlcvReplayDataSet#tick
```

About timeframe consolidation, you can look here
```java
OhlcvTimeframeConsolidation
// there are next examples of consolidations
RangeBarsIncrementalOhlcvConsolidation
VolumeIncrementalOhlcvConsolidation
```

Could you provide a hand-drawn example of the chart you're aiming for? I'm having trouble grasping the specifics of your goal from the description. Thanks.

**Reply by yezhengli-Mr9** [reply]
Dear @raven2cz , 
Thanks~ First goal is simple: real-time minute level candlesticks for intra day -- but for each candlestick want to put more points (moving average, some local extreme values).
<img width="1734" alt="DailyMarketStream 2023-11-20 07-32-27" src="https://github.com/fair-acc/chart-fx/assets/16505983/b1fd7c24-f8ee-49e9-8120-a33ac5e85554">
Feel "real-time“/handle with real-time streaming API is the only issue left for this first goal -- rest are resolved (I mean add extra points, etc.)
Following should help if indeed called in `loadTestdata(...)` and I will track:

https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/service/consolidate/IncrementalOhlcvConsolidation.java#L7
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/AbstractBasicFinancialApplication.java#L207-L215
as well as 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/service/SimpleOhlcvReplayDataSet.java#L202
Meanwhile, #378 may help.

**Reply by yezhengli-Mr9** [reply]
Can screenshot a current realized chart later today.

**Comment by yezhengli-Mr9** [comment]
# how to put code to get chart updated while `setData(...)` updates data timely already?

While consuming stream API in extended version of `FinancialRealtimeCandlestickSample`,can definitely update data similar as 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/AbstractBasicFinancialApplication.java#L319 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-dataset/src/main/java/io/fair_acc/dataset/spi/financial/OhlcvDataSet.java#L40-L42
in ways similar to  #378 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/chart/PaddedAutoGrowAxisSample.java#L40-L62
or 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/chart/Histogram2DimSample.java#L138-L144
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/chart/Histogram2DimSample.java#L39-L59

However, do not know **how to put code (likely between certain lines in the following) to get chart updated**? While tried to add following between lines:
```java
           CachedDaemonThreadFactory.getInstance().newThread(() -> {
            try {
                getStreamChartDataSetUpdate(symbolSignal, 1, null, 1, "Minute", true,
                        ohlcvDataSet,indiSet,
                        title, chart);
            }catch(IOException | ParseException e){
                e.printStackTrace();
            }
        }).start();
/**
  * getStreamChartDataSetUpdate(...)
  */
    // ...
             dataSet.setData(ohlcvOutput);// with new data
    //       Renderer renderer = new CandleStickWithMALocalMaxMinRenderer(true);
    //       renderer.getDatasets().addAll(ds);
    //       chart.getRenderers().clear();
    //       chart.getRenderers().add(renderer);
```
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/FinancialRealtimeCandlestickSample.java#L84-L152

**Comment by ennerf** [comment]
@yezhengli-Mr9  Can you provide more information about your application and what you are trying to achieve? That will make it easier to recommend something.

* Where is the data coming from?
* How latency sensitive is the application? Do you have specific targets?
* Are you planning on doing actual trading from that application?

You should also learn about the JavaFX threading model. [Architecture.md](https://github.com/fair-acc/chart-fx/blob/main/docs/Architecture.md) has a small summary about the individual phases. In general you can only update the chart from the JavaFX thread. Datasets have some built-in support for concurrent updates, but for low-latency applications with small updates it may make sense to do something more custom.

**Reply by yezhengli-Mr9** [reply]
Dear @ennerf ,
thanks
- API data feed, for example, similar to TradeStation API.
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/service/SimpleOhlcvDailyParser.java#L79-L81
- just updated every one minute. For short term, just need real-time (every one minute) candlestick update. 
At most to visualize some volume delta in the near future.
No need for any complicated stuff (no need for any button, etc.) But X-axis (time) with adaptive values are good.
-  just need to visualize real-time (per minute) GUI via chart-fx. Other non-GUI issue, can solve by java backend codes.

**Reply by ennerf** [reply]
Ok, that sounds fairly simple. Does it not work if you just add a new datapoint to the dataset? Changing the data should automatically trigger a chart update.

**Reply by yezhengli-Mr9** [reply]
Thanks, yes, just simple version. 
> Changing the data should automatically trigger a chart update.

Ok, that is the (main) thing I am checking from sample codes -- change the data, chart gets "updated automatically".

**Comment by yezhengli-Mr9** [comment]
Try to combine @raven2cz 's hints on `SimpleOhlcvReplayDataSet`, `de.gsi.financial.samples.service.SimpleOhlcvReplayDataSet#fillTestData` and simple `setData(newData)`, still no luck, attempts including:

1. add following 

```java
replayDataSet.addOhlcvChangeListener(ohlcvItem -> FXUtils.runFX(() -> {
                Interval<Calendar> timeRangeInt =null,ttInt=null;
                Calendar replayFromCal=null;
                try{
                    timeRangeInt = CalendarUtils.createByDateTimeInterval(timeRange);
                    ttInt = CalendarUtils.createByTimeInterval(tt);
                    replayFromCal = CalendarUtils.createByDateTime(replayFrom);
                }catch(ParseException pe){
                    pe.printStackTrace();
                }

                replayDataSet.fillTestData(period,
                        timeRangeInt,
                        ttInt, replayFromCal, consolidationAddons);
                closeIndicator.setMarkerValue(ohlcvItem.getClose());
            }));
```
instead of 
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/FinancialRealtimeCandlestickSample.java#L103
2.
```java
try {
                getStreamChartDataSetUpdate(symbolSignal, 1, null, 1, "Minute", true,
                        ohlcvDataSet,indiSet,
                        title, chart);
            }catch(IOException | ParseException e){
                e.printStackTrace();
            }
```
where within `getStreamChartDataSetUpdate(...)` there are
```java
BufferedReader responseReader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
                while ((line = responseReader.readLine()) != null) {

                    if (TokenResponse.aboutToExpire(tokenResponse, lagSeconds)) {
                        tokenResponse = Credential.getAccessToken();
                        continue;
                    }
                    if (!line.isEmpty()) {
                        // Assuming you have a JSONObject class for parsing JSON
                        Bar b = Bar.readBarFromJsonString(line, symbol);
                        System.out.println("[getStreamBarchartQuotes]" + b);
                        if (b == null) {
                            System.out.println("barList.size()\t" + barList.size() + ":\n" + barList.get(barList.size() - 1));
                            continue;
                        }
                        if (barList.size() == 0) {
                            barList.add(b);
                            // update the chart
                        } else {
                            long diff = Util.calculateTimeDiffSecUTC(b.getTimeStamp(), barList.get(barList.size() - 1).getTimeStamp());
                            if (diff > 30) {
                                barList.add(b);
                                /*
                                 * here is part to ohlcvDataSet.setData(...) new data
                                 */
                                Thread.sleep(20000); // try to sleep 20 seconds to provide chart enough time to update/refresh
                                // update the chart
                            } else {
                                barList.set(barList.size() - 1, b);
                            }
                        }
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
                System.out.println(connection.getResponseCode());
                System.out.println(connection.getErrorStream());
                System.out.println(connection.getInputStream());

            }
```
Will try more later today, tomorrow or later this week.
- should try more hints from @raven2cz ;
- reduce many unnecessary part of `FinancialRealtimeCandlestickSample`. **Only necessary part so far -- to see update chart, when adding one more point, or providing the chart new dataset by `setData(...)` .**
https://github.com/fair-acc/chart-fx/blob/ed86318c5a1bde23d285633a94860648aee5fb32/chartfx-samples/src/main/java/io/fair_acc/sample/financial/FinancialRealtimeCandlestickSample.java#L49
- Feel have to better understsand [Architecture.md](https://github.com/fair-acc/chart-fx/blob/main/docs/Architecture.md) and https://www.javatpoint.com/javafx-stacked-bar-chart, https://docs.oracle.com/javafx/2/charts/chart-overview.htm#CJAHHJCB

**Reply by ennerf** [reply]
ChartFX originally started off as an extension of the built-in Oracle charts, but over time it has become quite different. The content gets redrawn when a flag indicates that the corresponding parts are dirty.

It looks like `OhlcvDataSet::setData` does not do update notifications to trigger a redraw. You can manually trigger one using `dataSet.fireInvalidated(ChartBits.DataSetData);` after setting the data.

**Reply by yezhengli-Mr9** [reply]
Thanks very much, I noticed `fireInvalidated(ChartBits.DataSetDataAdded);` or `fireInvalidated(ChartBits.DataSetData);` earlier. Let me debug this part with more attention.

**Reply by ennerf** [reply]
The docs contain a section on the [event system](https://github.com/fair-acc/chart-fx/blob/main/docs/Architecture.md#event-system)

**Comment by yezhengli-Mr9** [comment]
Resolved.
